File: firmware/src/console/terminal.cpp
#include "terminal.h"

#include <string.h>

//------------------------------------------
//  Terminal
//------------------------------------------
console::Terminal::Terminal(char *buf, size_t cap)
    : buf_(buf), cap_(cap), len_(0), cursor_(0),
      last_char_(0), escape_state_(Normal) {
  buf_[0] = '\0';
}

console::KeyCode console::Terminal::process_byte(uint8_t byte) {
  switch (escape_state_) {
  case Normal:
    switch (byte) {
    case '\r': case '\n':     return KeyCode::Enter;
    case 0x01:               return KeyCode::CtrlA;
    case 0x02:               return KeyCode::CtrlB;
    case 0x03:               return KeyCode::CtrlC;
    case 0x04:               return KeyCode::CtrlD;
    case 0x05:               return KeyCode::CtrlE;
    case 0x06:               return KeyCode::CtrlF;
    case 0x08:               return KeyCode::Backspace;
    case 0x09:               return KeyCode::Tab;
    case 0x0B:               return KeyCode::CtrlK;
    case 0x0C:               return KeyCode::CtrlL;
    case 0x0E:               return KeyCode::CtrlN;
    case 0x10:               return KeyCode::CtrlP;
    case 0x12:               return KeyCode::CtrlR;
    case 0x14:               return KeyCode::CtrlT;
    case 0x15:               return KeyCode::CtrlU;
    case 0x17:               return KeyCode::CtrlW;
    case 0x1B:
      escape_state_ = Escape;
      return KeyCode::None;
    case 0x7F:               return KeyCode::Backspace;
    default:
      if (byte >= 0x20 && byte < 0x7F) {
        last_char_ = byte;
        return KeyCode::Char;
      }
      return KeyCode::None;
    }

  case Escape:
    if (byte == '[') {
      escape_state_ = Bracket;
      return KeyCode::None;
    }
    escape_state_ = Normal;
    return KeyCode::Escape;

  case Bracket:
    escape_state_ = Normal;
    switch (byte) {
    case 'A': return KeyCode::ArrowUp;
    case 'B': return KeyCode::ArrowDown;
    case 'C': return KeyCode::ArrowRight;
    case 'D': return KeyCode::ArrowLeft;
    case 'H': return KeyCode::Home;
    case 'F': return KeyCode::End;
    case '1': return KeyCode::Home;   // ESC[1~ (some terminals)
    case '3': return KeyCode::Delete; // ESC[3~
    case '4': return KeyCode::End;    // ESC[4~ (some terminals)
    default:  return KeyCode::None;
    }
  }

  return KeyCode::None;
}

console::TerminalEvent console::Terminal::handle_key(KeyCode key, uint8_t ch) {
  switch (key) {
  case KeyCode::Enter:
    return len_ == 0 ? TerminalEvent::EmptyCommand : TerminalEvent::CommandReady;

  case KeyCode::Backspace:
    if (cursor_ > 0 && len_ > 0) {
      memmove(buf_ + cursor_ - 1, buf_ + cursor_, len_ - cursor_);
      cursor_--;
      len_--;
      buf_[len_] = '\0';
      return TerminalEvent::BufferChanged;
    }
    return TerminalEvent::None;

  case KeyCode::Delete:
    if (cursor_ < len_) {
      memmove(buf_ + cursor_, buf_ + cursor_ + 1, len_ - cursor_ - 1);
      len_--;
      buf_[len_] = '\0';
      return TerminalEvent::BufferChanged;
    }
    return TerminalEvent::None;

  case KeyCode::ArrowLeft:
  case KeyCode::CtrlB:
    if (cursor_ > 0) {
      cursor_--;
      return TerminalEvent::CursorMoved;
    }
    return TerminalEvent::None;

  case KeyCode::ArrowRight:
  case KeyCode::CtrlF:
    if (cursor_ < len_) {
      cursor_++;
      return TerminalEvent::CursorMoved;
    }
    return TerminalEvent::None;

  case KeyCode::Home:
  case KeyCode::CtrlA:
    if (cursor_ > 0) {
      cursor_ = 0;
      return TerminalEvent::CursorHome;
    }
    return TerminalEvent::None;

  case KeyCode::End:
  case KeyCode::CtrlE:
    if (cursor_ < len_) {
      cursor_ = len_;
      return TerminalEvent::CursorEnd;
    }
    return TerminalEvent::None;

  case KeyCode::ArrowUp:
  case KeyCode::CtrlP:
    return TerminalEvent::HistoryPrevious;

  case KeyCode::ArrowDown:
  case KeyCode::CtrlN:
    return TerminalEvent::HistoryNext;

  case KeyCode::CtrlC:
    return TerminalEvent::Interrupt;

  case KeyCode::CtrlD:
    if (len_ == 0) return TerminalEvent::EndOfFile;
    if (cursor_ < len_) {
      memmove(buf_ + cursor_, buf_ + cursor_ + 1, len_ - cursor_ - 1);
      len_--;
      buf_[len_] = '\0';
      return TerminalEvent::BufferChanged;
    }
    return TerminalEvent::None;

  case KeyCode::CtrlK:
    if (cursor_ < len_) {
      len_ = cursor_;
      buf_[len_] = '\0';
      return TerminalEvent::KillToEnd;
    }
    return TerminalEvent::None;

  case KeyCode::CtrlL:
    return TerminalEvent::ClearScreen;

  case KeyCode::CtrlR:
    return TerminalEvent::Redraw;

  case KeyCode::CtrlT:
    if (cursor_ > 0 && cursor_ < len_) {
      char tmp = buf_[cursor_ - 1];
      buf_[cursor_ - 1] = buf_[cursor_];
      buf_[cursor_] = tmp;
      if (cursor_ < len_) cursor_++;
      return TerminalEvent::SwapChars;
    }
    return TerminalEvent::None;

  case KeyCode::CtrlW:
    return TerminalEvent::DeleteWord;

  case KeyCode::CtrlU:
    return TerminalEvent::ClearLine;

  case KeyCode::Char:
    if (len_ < cap_ - 1) {
      if (cursor_ < len_)
        memmove(buf_ + cursor_ + 1, buf_ + cursor_, len_ - cursor_);
      buf_[cursor_] = (char)last_char_;
      cursor_++;
      len_++;
      buf_[len_] = '\0';
      return TerminalEvent::BufferChanged;
    }
    return TerminalEvent::None;

  default:
    return TerminalEvent::None;
  }
}

const char *console::Terminal::buffer_str() const {
  return buf_;
}

size_t console::Terminal::buffer_length() const {
  return len_;
}

size_t console::Terminal::cursor_position() const {
  return cursor_;
}

void console::Terminal::set_buffer(const char *content) {
  size_t slen = strlen(content);
  if (slen >= cap_) slen = cap_ - 1;
  memcpy(buf_, content, slen);
  buf_[slen] = '\0';
  len_ = slen;
  cursor_ = slen;
}

void console::Terminal::clear_buffer() {
  len_ = 0;
  cursor_ = 0;
  buf_[0] = '\0';
}

const char *console::Terminal::take_command() {
  buf_[len_] = '\0';
  len_ = 0;
  cursor_ = 0;
  return buf_;
}

File: apps/ceratina/src/services/ssh/terminal.rs
use embassy_futures::select::{select, Either};
use embassy_sync::{blocking_mutex::raw::RawMutex, signal::Signal};
use embedded_io_async::{Read, Write as AsyncWrite};
use heapless::{String, Vec};

use super::history::History;
use super::writer::TerminalWriter;

/// Configuration for the terminal
#[derive(Clone, Copy)]
pub struct TerminalConfig {
    /// Maximum command buffer size
    pub buffer_size: usize,
    /// Prompt string to display
    pub prompt: &'static str,
    /// Enable echo of typed characters
    pub echo: bool,
    /// Enable ANSI escape codes for better terminal control
    pub ansi_enabled: bool,
}

impl Default for TerminalConfig {
    fn default() -> Self {
        Self {
            buffer_size: 128,
            prompt: "> ",
            echo: true,
            ansi_enabled: true,
        }
    }
}

/// Key codes for special keys
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum KeyCode {
    Backspace,
    Delete,
    Enter,
    Tab,
    Escape,
    ArrowUp,
    ArrowDown,
    ArrowLeft,
    ArrowRight,
    CtrlC,
    CtrlD,
    Char(u8),
}

/// Main terminal structure
pub struct Terminal<const BUF_SIZE: usize> {
    config: TerminalConfig,
    buffer: Vec<u8, BUF_SIZE>,
    cursor_pos: usize,
    escape_state: EscapeState,
}

/// State machine for parsing ANSI escape sequences
#[derive(Debug, Clone, Copy, PartialEq)]
enum EscapeState {
    Normal,
    Escape,
    Bracket,
}

impl<const BUF_SIZE: usize> Terminal<BUF_SIZE> {
    /// Create a new terminal instance
    pub fn new(config: TerminalConfig) -> Self {
        Self {
            config,
            buffer: Vec::new(),
            cursor_pos: 0,
            escape_state: EscapeState::Normal,
        }
    }

    /// Get the current buffer as a string slice
    pub fn buffer_str(&self) -> Result<&str, core::str::Utf8Error> {
        core::str::from_utf8(self.buffer.as_slice())
    }

    /// Clear the current buffer
    pub fn clear_buffer(&mut self) {
        self.buffer.clear();
        self.cursor_pos = 0;
    }

    /// Get the current cursor position
    pub fn cursor_position(&self) -> usize {
        self.cursor_pos
    }

    /// Process a single byte of input, handling ANSI escape sequences
    pub fn process_byte(&mut self, byte: u8) -> Option<KeyCode> {
        match self.escape_state {
            EscapeState::Normal => {
                match byte {
                    b'\r' | b'\n' => Some(KeyCode::Enter),
                    0x08 | 0x7F => Some(KeyCode::Backspace),
                    0x03 => Some(KeyCode::CtrlC),
                    0x04 => Some(KeyCode::CtrlD),
                    0x09 => Some(KeyCode::Tab),
                    0x1B => {
                        self.escape_state = EscapeState::Escape;
                        None
                    }
                    byte if byte >= 0x20 && byte < 0x7F => Some(KeyCode::Char(byte)),
                    _ => None,
                }
            }
            EscapeState::Escape => {
                if byte == b'[' {
                    self.escape_state = EscapeState::Bracket;
                    None
                } else {
                    self.escape_state = EscapeState::Normal;
                    Some(KeyCode::Escape)
                }
            }
            EscapeState::Bracket => {
                self.escape_state = EscapeState::Normal;
                match byte {
                    b'A' => Some(KeyCode::ArrowUp),
                    b'B' => Some(KeyCode::ArrowDown),
                    b'C' => Some(KeyCode::ArrowRight),
                    b'D' => Some(KeyCode::ArrowLeft),
                    b'3' => Some(KeyCode::Delete), // Delete sends ESC[3~
                    _ => None,
                }
            }
        }
    }

    /// Handle a key press
    pub fn handle_key(&mut self, key: KeyCode) -> TerminalEvent {
        match key {
            KeyCode::Enter => {
                if self.buffer.is_empty() {
                    TerminalEvent::EmptyCommand
                } else {
                    TerminalEvent::CommandReady
                }
            }
            KeyCode::Backspace => {
                if self.cursor_pos > 0 && !self.buffer.is_empty() {
                    self.buffer.remove(self.cursor_pos - 1);
                    self.cursor_pos -= 1;
                    TerminalEvent::BufferChanged
                } else {
                    TerminalEvent::None
                }
            }
            KeyCode::Delete => {
                if self.cursor_pos < self.buffer.len() {
                    self.buffer.remove(self.cursor_pos);
                    TerminalEvent::BufferChanged
                } else {
                    TerminalEvent::None
                }
            }
            KeyCode::ArrowLeft => {
                if self.cursor_pos > 0 {
                    self.cursor_pos -= 1;
                    TerminalEvent::CursorMoved
                } else {
                    TerminalEvent::None
                }
            }
            KeyCode::ArrowRight => {
                if self.cursor_pos < self.buffer.len() {
                    self.cursor_pos += 1;
                    TerminalEvent::CursorMoved
                } else {
                    TerminalEvent::None
                }
            }
            KeyCode::ArrowUp => TerminalEvent::HistoryPrevious,
            KeyCode::ArrowDown => TerminalEvent::HistoryNext,
            KeyCode::CtrlC => TerminalEvent::Interrupt,
            KeyCode::CtrlD => TerminalEvent::EndOfFile,
            KeyCode::Char(byte) => {
                if self.buffer.len() < BUF_SIZE {
                    // Insert at cursor position
                    if self.cursor_pos == self.buffer.len() {
                        let _ = self.buffer.push(byte);
                    } else {
                        let _ = self.buffer.insert(self.cursor_pos, byte);
                    }
                    self.cursor_pos += 1;
                    TerminalEvent::BufferChanged
                } else {
                    TerminalEvent::BufferFull
                }
            }
            _ => TerminalEvent::None,
        }
    }

    /// Get the current command buffer and clear it
    pub fn take_command(&mut self) -> Result<String<BUF_SIZE>, ()> {
        let result = String::from_utf8(self.buffer.clone()).map_err(|_| ())?;
        self.clear_buffer();
        Ok(result)
    }

    /// Set the buffer content (useful for history navigation)
    pub fn set_buffer(&mut self, content: &str) -> Result<(), ()> {
        self.buffer.clear();
        self.buffer.extend_from_slice(content.as_bytes()).map_err(|_| ())?;
        self.cursor_pos = self.buffer.len();
        Ok(())
    }
}

/// Events that can occur during terminal operation
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum TerminalEvent {
    None,
    BufferChanged,
    CursorMoved,
    CommandReady,
    EmptyCommand,
    BufferFull,
    Interrupt,
    EndOfFile,
    HistoryPrevious,
    HistoryNext,
}

/// Terminal reader task that handles async I/O
pub struct TerminalReader<const BUF_SIZE: usize> {
    terminal: Terminal<BUF_SIZE>,
    history: Option<History<BUF_SIZE>>,
}

impl<const BUF_SIZE: usize> TerminalReader<BUF_SIZE> {
    pub fn new(config: TerminalConfig, history: Option<History<BUF_SIZE>>) -> Self {
        Self {
            terminal: Terminal::new(config),
            history,
        }
    }

    /// Read a complete line from the input
    pub async fn read_line<R, W, M>(
        &mut self,
        reader: &mut R,
        writer: &mut TerminalWriter<'_, W>,
        redraw_signal: Option<&Signal<M, ()>>,
    ) -> Result<String<BUF_SIZE>, ReadLineError>
    where
        R: Read,
        W: AsyncWrite,
        M: RawMutex,
    {
        // Display initial prompt
        let _ = writer.write_prompt(self.terminal.config.prompt).await;

        let mut byte_buf = [0u8; 1];

        loop {
            let event = if let Some(signal) = redraw_signal {
                // Wait for either input or redraw signal
                match select(reader.read(&mut byte_buf), signal.wait()).await {
                    Either::First(Ok(1)) => {
                        if let Some(key) = self.terminal.process_byte(byte_buf[0]) {
                            self.terminal.handle_key(key)
                        } else {
                            TerminalEvent::None
                        }
                    }
                    Either::First(Ok(0)) => TerminalEvent::EndOfFile,
                    Either::First(Err(_)) => return Err(ReadLineError::IoError),
                    Either::Second(_) => {
                        // Redraw requested
                        signal.reset();
                        let _ = writer.clear_line().await;
                        let _ = writer.write_prompt(self.terminal.config.prompt).await;
                        let _ = writer.write_str(self.terminal.buffer_str().unwrap_or("")).await;
                        continue;
                    }
                    _ => continue,
                }
            } else {
                // Simple read without redraw support
                match reader.read(&mut byte_buf).await {
                    Ok(1) => {
                        if let Some(key) = self.terminal.process_byte(byte_buf[0]) {
                            self.terminal.handle_key(key)
                        } else {
                            TerminalEvent::None
                        }
                    }
                    Ok(0) => TerminalEvent::EndOfFile,
                    Err(_) => return Err(ReadLineError::IoError),
                    _ => continue,
                }
            };

            match event {
                TerminalEvent::CommandReady => {
                    let _ = writer.write_str("\r\n").await;
                    let command = self.terminal.take_command()?;
                    
                    // Add to history if available
                    if let Some(ref mut hist) = self.history {
                        let _ = hist.add(&command);
                    }
                    
                    return Ok(command);
                }
                TerminalEvent::EmptyCommand => {
                    let _ = writer.write_str("\r\n").await;
                    let _ = writer.write_prompt(self.terminal.config.prompt).await;
                }
                TerminalEvent::BufferChanged => {
                    if self.terminal.config.echo {
                        // Redraw the line
                        let _ = writer.clear_line().await;
                        let _ = writer.write_prompt(self.terminal.config.prompt).await;
                        let _ = writer.write_str(self.terminal.buffer_str().unwrap_or("")).await;
                    }
                }
                TerminalEvent::Interrupt => {
                    self.terminal.clear_buffer();
                    let _ = writer.write_str("^C\r\n").await;
                    let _ = writer.write_prompt(self.terminal.config.prompt).await;
                }
                TerminalEvent::EndOfFile => {
                    return Err(ReadLineError::EndOfFile);
                }
                TerminalEvent::HistoryPrevious => {
                    if let Some(ref mut hist) = self.history {
                        if let Some(entry) = hist.previous() {
                            let _ = self.terminal.set_buffer(entry);
                            // Redraw the line
                            let _ = writer.clear_line().await;
                            let _ = writer.write_prompt(self.terminal.config.prompt).await;
                            let _ = writer.write_str(self.terminal.buffer_str().unwrap_or("")).await;
                        }
                    }
                }
                TerminalEvent::HistoryNext => {
                    if let Some(ref mut hist) = self.history {
                        if let Some(entry) = hist.next() {
                            let _ = self.terminal.set_buffer(entry);
                        } else {
                            // At the end of history, clear buffer
                            self.terminal.clear_buffer();
                        }
                        // Redraw the line
                        let _ = writer.clear_line().await;
                        let _ = writer.write_prompt(self.terminal.config.prompt).await;
                        let _ = writer.write_str(self.terminal.buffer_str().unwrap_or("")).await;
                    }
                }
                TerminalEvent::BufferFull => {
                    // Optionally signal buffer full (beep?)
                }
                _ => {}
            }
        }
    }
}

/// Errors that can occur while reading a line
#[derive(Debug, Clone, Copy)]
pub enum ReadLineError {
    IoError,
    Utf8Error,
    EndOfFile,
}

impl From<()> for ReadLineError {
    fn from(_: ()) -> Self {
        ReadLineError::Utf8Error
    }
}
